In [3]:
import pandas as pd
import numpy as np
import json
import os
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
from sklearn.preprocessing import OneHotEncoder
from mlxtend.preprocessing import OnehotTransactions

In [4]:
# Load in the data
with open('Assoc_Rule_DataF_707d.json','r') as file:
        data = json.load(file)

In [5]:
test = data
test[0]

[[{'hero_id': 19,
   'hero_name': ['Tiny'],
   'is_pick': False,
   'match_id': 3626698786,
   'order': 0,
   'team': 1},
  {'hero_id': 107,
   'hero_name': ['Earth Spirit'],
   'is_pick': False,
   'match_id': 3626698786,
   'order': 1,
   'team': 0},
  {'hero_id': 60,
   'hero_name': ['Night Stalker'],
   'is_pick': False,
   'match_id': 3626698786,
   'order': 2,
   'team': 1},
  {'hero_id': 79,
   'hero_name': ['Shadow Demon'],
   'is_pick': False,
   'match_id': 3626698786,
   'order': 3,
   'team': 0},
  {'hero_id': 85,
   'hero_name': ['Undying'],
   'is_pick': False,
   'match_id': 3626698786,
   'order': 4,
   'team': 1},
  {'hero_id': 3,
   'hero_name': ['Bane'],
   'is_pick': False,
   'match_id': 3626698786,
   'order': 5,
   'team': 0},
  {'hero_id': 68,
   'hero_name': ['Ancient Apparition'],
   'is_pick': True,
   'match_id': 3626698786,
   'order': 6,
   'team': 1},
  {'hero_id': 16,
   'hero_name': ['Sand King'],
   'is_pick': True,
   'match_id': 3626698786,
   'order

In [6]:
list = []
tempList = []
# Coding is: Order_Pick(T/F)_Hero_Team(0/1)
for i in range(len(test)):
    for x in range(len(test[i][0])):
        tempList.append('{}{}{}{}'.format(test[i][0][x]['order'],test[i][0][x]['is_pick'],test[i][0][x]['hero_name'],test[i][0][x]['team']))
        if test[i][1] == True:
            tempList.append('Rad_Win')
        else:
            tempList.append('Dire_Win')
            
    list.append(tempList)
    tempList = []

Do One Hot Encoding on the data so that we can run the apriori algorithm

In [12]:
oht = OnehotTransactions()
oht_ary = oht.fit(list).transform(list)
df = pd.DataFrame(oht_ary, columns=oht.columns_)

C:\Users\lizar\Anaconda2\envs\py3k\lib\site-packages\mlxtend\preprocessing\onehot.py:66: DeprecationWarning: OnehotTransactions has been deprecated and will be removed in future. Please use TransactionEncoder instead.
  warnings.warn(msg, DeprecationWarning)


Because there are a lot of different heros and draft pick orders all of the observations will have quite low support. However, we can still see which ones occur the most often.

In [13]:
apriori(df,min_support=.05,use_colnames = True,max_len = 2)

support                    itemsets
0   0.058621  [0False['Bounty Hunter']1]
1   0.075000      [0False['Clockwerk']1]
2   0.062931  [0False['Night Stalker']1]
3   0.050862           [1False['Bane']0]
4   0.071552  [1False['Bounty Hunter']0]
5   0.081034  [1False['Night Stalker']0]
6   0.050000  [1False['Night Stalker']1]
7   0.073276     [1False['Omniknight']0]
8   0.065517      [2False['Clockwerk']1]
9   0.062931     [2False['Omniknight']1]
10  0.068103  [3False['Bounty Hunter']0]
11  0.051724      [3False['Clockwerk']0]
12  0.068966  [3False['Night Stalker']0]
13  0.051724  [3False['Night Stalker']1]
14  0.057759     [3False['Omniknight']0]
15  0.061207     [4False['Omniknight']1]
16  0.050862           [5False['Bane']0]
17  0.050862  [5False['Night Stalker']0]
18  0.070690     [5False['Omniknight']0]
19  0.056034            [6True['Bane']1]
20  0.064655       [6True['Clockwerk']1]
21  0.050000   [6True['Night Stalker']1]
22  0.056034            [6True['Tiny']1]
23  0.056897            [6True['Tusk']0]
24  0.080172            [6True['Tusk']1]
25  0.054310    [8True['Shadow Fiend']0]
26  0.494828                  [Dire_Win]
27  0.505172                   [Rad_Win]

What occurs the most often?

In [14]:
frequent_itemsets = apriori(df, min_support=0.03, use_colnames=True)
frequent_itemsets['length'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
frequent_itemsets

support                              itemsets  length
0    0.033621                     [0False['Bane']1]       1
1    0.058621            [0False['Bounty Hunter']1]       1
2    0.031897                [0False['Clockwerk']0]       1
3    0.075000                [0False['Clockwerk']1]       1
4    0.043966            [0False['Night Stalker']0]       1
5    0.062931            [0False['Night Stalker']1]       1
6    0.042241               [0False['Omniknight']0]       1
7    0.038793               [0False['Omniknight']1]       1
8    0.038793                     [0False['Tiny']1]       1
9    0.034483                     [0False['Tusk']1]       1
10   0.033621                   [11False['Lycan']1]       1
11   0.035345                   [11False['Razor']1]       1
12   0.031034                   [12False['Razor']0]       1
13   0.032759                   [13False['Lycan']1]       1
14   0.036207                   [13False['Razor']1]       1
15   0.032759              [13False['Tidehunter']1]       1
16   0.032759                [14True['Disruptor']0]       1
17   0.040517                [15True['Disruptor']1]       1
18   0.040517           [18False['Death Prophet']0]       1
19   0.031034                  [18False['Tinker']0]       1
20   0.037069          [19False['Phantom Lancer']1]       1
21   0.034483                  [19False['Tinker']1]       1
22   0.050862                     [1False['Bane']0]       1
23   0.071552            [1False['Bounty Hunter']0]       1
24   0.034483            [1False['Bounty Hunter']1]       1
25   0.044828                [1False['Clockwerk']0]       1
26   0.031034                [1False['Clockwerk']1]       1
27   0.081034            [1False['Night Stalker']0]       1
28   0.050000            [1False['Night Stalker']1]       1
29   0.073276               [1False['Omniknight']0]       1
..        ...                                   ...     ...
106  0.033621       [0False['Clockwerk']1, Rad_Win]       2
107  0.035345  [0False['Night Stalker']1, Dire_Win]       2
108  0.032759  [1False['Bounty Hunter']0, Dire_Win]       2
109  0.038793   [1False['Bounty Hunter']0, Rad_Win]       2
110  0.033621  [1False['Night Stalker']0, Dire_Win]       2
111  0.047414   [1False['Night Stalker']0, Rad_Win]       2
112  0.035345     [1False['Omniknight']0, Dire_Win]       2
113  0.037931      [1False['Omniknight']0, Rad_Win]       2
114  0.031897       [2False['Clockwerk']0, Rad_Win]       2
115  0.031034      [2False['Clockwerk']1, Dire_Win]       2
116  0.034483       [2False['Clockwerk']1, Rad_Win]       2
117  0.033621     [2False['Omniknight']1, Dire_Win]       2
118  0.036207  [3False['Bounty Hunter']0, Dire_Win]       2
119  0.031897   [3False['Bounty Hunter']0, Rad_Win]       2
120  0.033621  [3False['Night Stalker']0, Dire_Win]       2
121  0.035345   [3False['Night Stalker']0, Rad_Win]       2
122  0.031034   [3False['Night Stalker']1, Rad_Win]       2
123  0.030172      [3False['Omniknight']0, Rad_Win]       2
124  0.031897      [4False['Omniknight']1, Rad_Win]       2
125  0.031034  [5False['Night Stalker']0, Dire_Win]       2
126  0.037931     [5False['Omniknight']0, Dire_Win]       2
127  0.032759      [5False['Omniknight']0, Rad_Win]       2
128  0.030172            [6True['Bane']1, Dire_Win]       2
129  0.032759       [6True['Clockwerk']1, Dire_Win]       2
130  0.031897        [6True['Clockwerk']1, Rad_Win]       2
131  0.034483   [6True['Night Stalker']1, Dire_Win]       2
132  0.032759            [6True['Tusk']0, Dire_Win]       2
133  0.038793            [6True['Tusk']1, Dire_Win]       2
134  0.041379             [6True['Tusk']1, Rad_Win]       2
135  0.032759     [8True['Shadow Fiend']0, Rad_Win]       2

[136 rows x 3 columns]

Lets filter our results to show some associations between features by setting 'length' to 2 and sorting the results.
Now we can see if certain picks at certain times during the draft cause teams to win more often.

In [10]:
frequent_itemsets[ (frequent_itemsets['length'] == 2) ].sort_values('support',ascending=False)

support                              itemsets  length
111  0.047414   [1False['Night Stalker']0, Rad_Win]       2
134  0.041379             [6True['Tusk']1, Rad_Win]       2
105  0.041379      [0False['Clockwerk']1, Dire_Win]       2
133  0.038793            [6True['Tusk']1, Dire_Win]       2
109  0.038793   [1False['Bounty Hunter']0, Rad_Win]       2
113  0.037931      [1False['Omniknight']0, Rad_Win]       2
126  0.037931     [5False['Omniknight']0, Dire_Win]       2
118  0.036207  [3False['Bounty Hunter']0, Dire_Win]       2
121  0.035345   [3False['Night Stalker']0, Rad_Win]       2
107  0.035345  [0False['Night Stalker']1, Dire_Win]       2
112  0.035345     [1False['Omniknight']0, Dire_Win]       2
131  0.034483   [6True['Night Stalker']1, Dire_Win]       2
116  0.034483       [2False['Clockwerk']1, Rad_Win]       2
120  0.033621  [3False['Night Stalker']0, Dire_Win]       2
117  0.033621     [2False['Omniknight']1, Dire_Win]       2
110  0.033621  [1False['Night Stalker']0, Dire_Win]       2
106  0.033621       [0False['Clockwerk']1, Rad_Win]       2
132  0.032759            [6True['Tusk']0, Dire_Win]       2
129  0.032759       [6True['Clockwerk']1, Dire_Win]       2
127  0.032759      [5False['Omniknight']0, Rad_Win]       2
135  0.032759     [8True['Shadow Fiend']0, Rad_Win]       2
108  0.032759  [1False['Bounty Hunter']0, Dire_Win]       2
124  0.031897      [4False['Omniknight']1, Rad_Win]       2
130  0.031897        [6True['Clockwerk']1, Rad_Win]       2
119  0.031897   [3False['Bounty Hunter']0, Rad_Win]       2
114  0.031897       [2False['Clockwerk']0, Rad_Win]       2
125  0.031034  [5False['Night Stalker']0, Dire_Win]       2
122  0.031034   [3False['Night Stalker']1, Rad_Win]       2
115  0.031034      [2False['Clockwerk']1, Dire_Win]       2
104  0.031034   [0False['Bounty Hunter']1, Rad_Win]       2
123  0.030172      [3False['Omniknight']0, Rad_Win]       2
128  0.030172            [6True['Bane']1, Dire_Win]       2

It looks like the top result is that Radiant wins 4.7% of the time when they ban Night Stalker first. This can make sense as this data comes from pub matches and during patch 7.07d Night Stalker was quite strong in pub matches.